contributors/design-proposals/network/pod-resolv-conf.md

```diff
@@ -0,0 +1,224 @@
+
+The current work around for this problem is to specify an FQDN for name
+resolution. Any domain name that ends with a period (e.g. `foo.bar.com.`) will
+not be search path expanded. However, use of FQDNs is not well-known practice
+and imposes application-level changes. Cluster operators may not have the luxury
+of enforcing such a change to applications that run on their infrastructure.
+
+It is also possible for the user to insert a short shell script snippet that
+rewrites `resolv.conf` on container start-up. This has the same problems as the
+previous approach and is also awkward for the user. This also forces the
+container to have additional executable code such as a shell or scripting engine
+which increases the applications security surface area.
+
+
+# Proposal sketch
+
+Add a new `pod.dnsPolicy: Custom` that allows for user customization of
+`resolv.conf`.
+
+
+## Pod API example
+
+In the example below, the user wishes to add the pod namespace and a custom
+expansion to the search path, as they do not use the other name aliases:
+
+```yaml
+# Pod spec
+apiVersion: v1
+kind: Pod
+metadata:
+  namespace: ns1
+  name: example
+spec:
+  containers:
+  - name: example
+    image: example
+  dnsPolicy: Custom
+  dnsParams:
+    custom:
+      search:
+      - $(NAMESPACE).svc.$(CLUSTER)
+      - my.dns.search.suffix
+      - $(HOST)
+      options:
+      - "ndots:2"
+```
+
+Given the following host `/etc/resolv.conf`:
+
+```bash
+nameserver 1.2.3.4
+search foo.com bar.com
+options ndots:1
+```
+
+The pod will get the following `/etc/resolv.conf`:
+
+```bash
+nameserver 10.240.0.10
+# Comments only for explication purposes.
+#
+#      $(NAMESPACE).svc.$CLUSTER
+#      |                     my.dns.search.suffix
+#      |                     |                    $(HOST)
+#      |                     |                    |
+#      V                     V                    V
+search ns1.svc.cluster.local my.dns.search.suffix foo.com bar.com
+# Populated from custom.options
+options ndots:2
+```
+
+## More examples
+
+The following is a Pod dnsParams that only contains the host search paths:
+
+```yaml
+dnsParams:
+  custom:
+    searchPaths:
+    - $HOST
+```
+
+Override `ndots` and add custom search path. Note that overriding the ndot may
+break the functionality of some of the search paths the
+
+```yaml
+dnsParams:
+  custom:
+    searchPaths:
+    - my.custom.suffix
+    - $HOST
+    options:
+    - "ndots:3"
+```
+
+# API changes
+
+```go
+type PodSpec struct {
+	...
+	DNSPolicy string
+	DNSParams *PodDNSParams
+    ...
+}
+
+type PodDNSParams struct {
+ 	Custom PodDNSParamsCustom
+}
+
+type PodDNSParamsCustom struct {
+	Search []string
+ 	Options []string
+}
+
+// This will not appear in types.go but is here for explication purposes.
+type DNSParamsSubstitution string
+const (
+	DNSParamsSearchPathNamespace = "$(NAMESPACE)"
+	DNSParamsSearchPathClusterDomain = "$(CLUSTER)"
+	DNSParamsSearchPathHostPaths = "$(HOST)"
```

Comment:
$(HOST) feels wrong.  Can we find a better name?  $(HOST_SEARCHES) or something?